src/framework/linalg/eigen_psd.hpp

```diff
@@ -22,17 +22,168 @@
  * Returns the eigenvalues and eigenvectors of a Hermitian
  * positive semi-definite (PSD) matrix.
  * @param psd_matrix: The Hermitian PSD matrix.
- * @param eigenvalues: The eignevalues of the matrix.
- * @param eigenvectors: The eigenvectors of the matrix.
+ * @param eigenvalues: On output: the eignevalues of the matrix (input is overwritten)
+ * @param eigenvectors: On input: a copy of psd_matrix, on output: the eigenvectors of the matrix.
  *
- * @returns: true.
+ * @returns: void
  */
-template <class float_t>
-void eigensystem_psd(const matrix<std::complex<float_t>>& psd_matrix,
-               /* out */ std::vector<std::complex<float_t>>& eigenvalues,
-               /* out */ std::vector<std::vector<std::complex<float_t>>>& eigenvectors){
+template <class T>
+void eigensystem_psd(const matrix<std::complex<T>>& psd_matrix,
+               /* out */ std::vector<T>& eigenvalues,
+               /* out */ matrix<std::complex<T>>& eigenvectors);
 
+//template <>
+//void eigensystem_psd(const matrix<std::complex<float>>& psd_matrix,
+//               /* out */ std::vector<float>& eigenvalues,
+//               /* out */ matrix<std::complex<float>>& eigenvectors); 
+//template <>
+//void eigensystem_psd(const matrix<std::complex<double>>& psd_matrix,
+//               /* out */ std::vector<double>& eigenvalues,
+//               /* out */ matrix<std::complex<double>>& eigenvectors); 
 
+template <class T>
+void eigensystem_psd(const matrix<std::complex<T>>& psd_matrix,
+               /* out */ std::vector<T>& eigenvalues,
+               /* out */ matrix<std::complex<T>>& eigenvectors) {}
+
+template <>
+void eigensystem_psd(const matrix<std::complex<double>>& psd_matrix,
+               /* out */ std::vector<double>& eigenvalues,
+               /* out */ matrix<std::complex<double>>& eigenvectors) {
+#ifdef DEBUG
+  if ( psd_matrix.getRows() != psd_matrix.getCols() ) {
+    std::cerr
```

Comment:
(single reply to other debug logging related comments) - I'll definitely take a look at the pulse logging mech, thanks for the pointer - I just did this this way because I had seen it done other places nearby and I hadn't heard of any more sophisticated logging functionality elsewhere. Depending on the compiler/platform though, using cout can be a very performance intensive op - I'd bet on macOS it's pretty efficient - I know there are better ways of doing it, but didn't want to implement one myself here, will update though